File: FashionSenseOutfits/FashionSenseOutfits.cs
// Copyright (C) 2023 Nihilistzsche
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.



namespace FashionSenseOutfits
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    using FashionSense.Framework.Interfaces.API;
    using ContentPatcher;

    using Models;

    using StardewModdingAPI;
    using StardewModdingAPI.Events;

    using StardewValley;

    using OutfitDataModel = System.Collections.Generic.Dictionary<string, Models.OutfitData>;

    /// <summary>The mod entry point.</summary>
    // ReSharper disable once UnusedMember.Global
    public class FashionSenseOutfits : Mod
    {
        private const string AssetName = "nihilistzsche.FashionSenseOutfits/Outfits";
        private static IApi _fsApi;
        private static IContentPatcherAPI _cpApi;
        private static OutfitDataModel _data;

        public override void Entry(IModHelper helper)
        {
            helper.Events.GameLoop.GameLaunched += OnGameLaunched;
            helper.Events.GameLoop.SaveLoaded += OnSaveLoaded;
            helper.Events.GameLoop.DayStarted += OnDayStarted;
            helper.Events.GameLoop.TimeChanged += OnTimeChanged;
            helper.Events.Player.Warped += OnWarped;
            helper.Events.Content.AssetRequested += OnAssetRequested;
            helper.Events.Content.AssetReady += OnAssetReady;
        }

        private static (bool Valid, string CorrectedID) IsValid(string outfitId)
        {
            if (string.IsNullOrEmpty(outfitId))
            {
                return (false, null);
            }

            var outfitIDs = _fsApi.GetOutfitIds().Value;
            var correctedId = outfitIDs.FirstOrDefault(x => x.Equals(outfitId, StringComparison.OrdinalIgnoreCase));
            return (correctedId != null, correctedId);
        }

        private void OnGameLaunched(object sender, GameLaunchedEventArgs e)
        {
            _fsApi = Helper.ModRegistry.GetApi<IApi>("PeacefulEnd.FashionSense");
            _cpApi = Helper.ModRegistry.GetApi<IContentPatcherAPI>("Pathoschild.ContentPatcher");
            _cpApi.RegisterToken(this.ModManifest, "CurrentOutfit", () =>
            {
                if (!Context.IsWorldReady)
                {
                    return null;
                }

                var outfitPair = _fsApi.GetCurrentOutfitId();
                return new[] { outfitPair.Key ? outfitPair.Value : null };
            });
        }

        private void LoadData(dynamic e)
        {
            // ReSharper disable once InvertIf
            if (e.GetType().GetProperty("IsLocalPlayer") == null || e.IsLocalPlayer)
            {
                Helper.GameContent.InvalidateCache(AssetName);
                _data = Game1.content.Load<OutfitDataModel>(AssetName);
            }
        }

        private void OnSaveLoaded(object sender, SaveLoadedEventArgs e) => LoadData(e);

        private void OnDayStarted(object sender, DayStartedEventArgs e) => LoadData(e);

        private void OnWarped(object sender, WarpedEventArgs e) => LoadData(e);

        private void OnTimeChanged(object sender, TimeChangedEventArgs e)
        {
            if (e.NewTime % 100 == 0)
            {
                LoadData(e);
            }
        }

        private static readonly OutfitDataModel BaseData = new()
        {
            ["CurrentOutfit"] = new OutfitData { OutfitId = string.Empty },
        };

        // ReSharper disable once MemberCanBeMadeStatic.Local
        private void OnAssetRequested(object sender, AssetRequestedEventArgs e)
        {
            if (!e.Name.IsEquivalentTo(AssetName)) return;
            e.LoadFrom(
                () => BaseData,
                AssetLoadPriority.Medium);
        }

        private string RequestedOutfitId => _data["CurrentOutfit"].OutfitId;

        private void OnAssetReady(object sender, AssetReadyEventArgs e)
        {
            if (!e.Name.IsEquivalentTo(AssetName)) return;
            var requestedOutfitId = RequestedOutfitId;
            var currentOutfitPair = _fsApi.GetCurrentOutfitId();
            var (valid, correctedId) = IsValid(requestedOutfitId);
            if (!valid)
            {
                Monitor.Log($"Given outfit with ID {requestedOutfitId} is invalid.");
                return;
            }

            if (currentOutfitPair.Key && correctedId == currentOutfitPair.Value)
            {
                Monitor.Log($"Skipping because the outfit with ID {correctedId} is already equipped.");
            }
            else
            {
                Monitor.Log($"Applying outfit with ID {correctedId} via Fashion Sense API...");
                _fsApi.SetCurrentOutfitId(correctedId, ModManifest);
            }
        }
    }
}

File: FashionSenseOutfits/FashionSenseOutfits.fs
// Copyright (C) 2023 Nihilistzsche
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

namespace FashionSenseOutfits

open System
open System.Collections.Generic
open System.Linq
open Microsoft.FSharp.Collections

open FashionSense.Framework.Interfaces.API
open ContentPatcher

open StardewModdingAPI
open StardewModdingAPI.Events

open StardewValley

open FashionSenseOutfits.Models

type private OutfitDataModel = Dictionary<string, OutfitData>

type public FashionSenseOutfits() =
    inherit Mod()

    static let __assetName: string = "nihilistzsche.FashionSenseOutfits/Outfits"
    static let mutable __fsApi: IApi = null
    static let mutable __cpApi: IContentPatcherAPI = null
    member val private _data: OutfitDataModel = null with get, set 
    member val private _cpConditionsReady = false with get, set
    member val private _lastEvent: Event = null with get, set
    member val private _seenInvalids = [] with get, set
    
    member private _.IsValid(requestedOutfitId: string): bool*string =
        if String.IsNullOrEmpty(requestedOutfitId) then
            (false, null)
        else
            let correctedId = __fsApi.GetOutfitIds().Value.FirstOrDefault(fun outfitId -> outfitId.Equals(requestedOutfitId, StringComparison.OrdinalIgnoreCase))
            (not (isNull correctedId), correctedId)

    member private this.RequestData(e: obj) =
        let isLocal = e.GetType().GetProperty("IsLocalPlayer")
        this.Monitor.Log($"Requesting data with args {e.GetType()}")
        if isNull isLocal || isLocal.GetValue(e) :?> bool then
            Game1.content.Load<OutfitDataModel>(__assetName) |> ignore
    
    member inline private _.ValidateAsset<'T when 'T : (member Name : IAssetName)>(e: 'T) =
        e.Name.IsEquivalentTo(__assetName)

    member private this.UpdateOutfit() =
        let requestedOutfitId = this._data["RequestedOutfit"].OutfitId
        let valid, correctedId = this.IsValid(requestedOutfitId)
        this.Monitor.Log($"Checking outfit {requestedOutfitId}")
        if valid then
            let currentOutfitPair = __fsApi.GetCurrentOutfitId()
            if not currentOutfitPair.Key || correctedId <> currentOutfitPair.Value then
                __fsApi.SetCurrentOutfitId(correctedId, this.ModManifest) |> ignore
        else
            if not (List.exists(fun elem -> elem = requestedOutfitId) this._seenInvalids) && requestedOutfitId <> "" then
                this._seenInvalids <- requestedOutfitId :: this._seenInvalids
                this.Monitor.Log($"Given outfit with ID {requestedOutfitId} is invalid.")

    member private this.OnGameLaunched(_: GameLaunchedEventArgs) =
        __fsApi <- this.Helper.ModRegistry.GetApi<IApi>("PeacefulEnd.FashionSense")
        __cpApi <- this.Helper.ModRegistry.GetApi<IContentPatcherAPI>("Pathoschild.ContentPatcher")
        __cpApi.RegisterToken(this.ModManifest, 
            "CurrentOutfit", 
            Func<string[]>(fun () ->
                if not Context.IsWorldReady then
                    [| null |]
                else
                    let outfitPair = __fsApi.GetCurrentOutfitId()
                    [| if outfitPair.Key then outfitPair.Value else null |]
            )
        )
    
    member private this.OnUpdateTicked(e: UpdateTickedEventArgs) =
        if (not (isNull this._lastEvent) && isNull Game1.CurrentEvent) || (not this._cpConditionsReady && __cpApi.IsConditionsApiReady) then
            if not this._cpConditionsReady then this._cpConditionsReady <- __cpApi.IsConditionsApiReady
            this.RequestData(e)
        this._lastEvent <- Game1.CurrentEvent

    member inline private _.KVP<'T,'U>(k: 'T, v: 'U) =
        KeyValuePair<'T,'U>(k, v)

    member private this.LoadModel: unit -> obj =
        fun() -> OutfitDataModel([ this.KVP("RequestedOutfit", OutfitData(String.Empty)) ])
    
    member private this.OnAssetRequested(e: AssetRequestedEventArgs) =
        if this.ValidateAsset(e) then
            e.LoadFrom(this.LoadModel, AssetLoadPriority.Medium)

    member private this.OnAssetReady(e: AssetReadyEventArgs) =
        if this.ValidateAsset(e) then
            let newData = Game1.content.Load<OutfitDataModel>(__assetName)
            if isNull(this._data) || this._data["RequestedOutfit"].OutfitId <> newData["RequestedOutfit"].OutfitId then
                this._data <- newData
                this.UpdateOutfit()
                
    override this.Entry(helper: IModHelper) =
        helper.Events.GameLoop.GameLaunched.Add(this.OnGameLaunched)
        helper.Events.GameLoop.DayStarted.Add(this.RequestData)
        helper.Events.GameLoop.TimeChanged.Add(this.RequestData)
        helper.Events.GameLoop.UpdateTicked.Add(this.OnUpdateTicked)
        helper.Events.Player.Warped.Add(this.RequestData)
        helper.Events.Content.AssetRequested.Add(this.OnAssetRequested)
        helper.Events.Content.AssetReady.Add(this.OnAssetReady)
